### ISSUE TITLE
lexical scoping elisp?

### ISSUE THREAD

**Issue by andreyorst** [greghendershott/racket-mode#441]
I've started using the Emacs [`nativecomp`](https://akrl.sdf.org/gccemacs.html) branch, which provides a feature to compile Emacs Lisp into native code. Currently`gccemacs` can't compile non-lexically scoped `.el` files:

> A lisp file can be compiled as follows:
>
>```emacs-lisp
>(native-compile-async "file.el")
>```
>Once the compilation completes, the produced file.eln can be loaded conventionally.
>**Please note that file.el has to be lexically bound in order to be compiled.**

Since I'm using this branch, I look at compilation buffer, and see some errors, that I'm trying to report where possible.

The thing is that I've thought that `lexical-binding` is pretty much first thing to set when writing new Emacs package, so not setting it may be because it was forgotten, or dynamic scope is necessary, which I've never seen actually. I think it is not set by default, because too much pre Emacs 24 packages will break, but new ones supposed to be using it. I've checked that some files in racket-mode set `lexical-binding` to `t`, and some are not. Is this intentional?

**Comment by greghendershott** [comment]
Long story: When I started Racket Mode nearly 10 years ago, `lexical-binding` either didn't exist or at least wasn't yet widely used.

For a long time, it seemed (to me) just optional -- use it if you wanted to write code that truly needed a lexical closure to work correctly, at all.

In recent years, it's become clearer (to me) that it's good to use, always, because [it can be faster](https://nullprogram.com/blog/2016/12/22/).

---

So far, my approach has been to add this to Racket Mode `.el` file if/as/when I'm updating them for some other reason, and have an opportunity to review them. I've been concerned that there's a non-zero probability that just adding it, wholesale, might result in some subtle bug. So I haven't done that, so far.

---

TL;DR: I'd be glad to add this to the remaining files. I'd still prefer to do it deliberately, one by one. However I could pick up the pace.

**Comment by andreyorst** [comment]
> TL;DR: I'd be glad to add this to the remaining files. I'd still prefer to do it deliberately, one by one. However I could pick up the pace.

No need to hurry. Native compilation is just a feature branch right now, so it's not the time to adopt everything.

I understood the situation, so I think I can manually add those flags locally and see if anything breaks, then report here. Will that help?

**Comment by greghendershott** [comment]
> I understood the situation, so I think I can manually add those flags locally and see if anything breaks, then report here. Will that help?

That would be great -- thanks!

**Comment by andreyorst** [comment]
Oh, I had OOM  while native compiling some files by simply adding `lexical-scope` to first line. But I see that you've changed some code, perhaps that was the reason. (my report was not yet ready though since I was invesitgating OOM cases)

**Comment by greghendershott** [comment]
I didn't mean to catch you off guard. I had more time to consider the remaining files, and didn't see any deal-breakers for adding it.

Having said that, if there are OOM native compile errors for any files, I'd be happy to take a look at that (or happy if you wanted to take a look)!

**Comment by andreyorst** [comment]
>  I had more time to consider the remaining files, and didn't see any deal-breakers for adding it.

Yes, I see that you've added this about 10 days ago, and just committed it upstream yesterday. No it seems there's no issues with OOM so far, and everything compiles and works correctly. Big thanks! REPL starts mostly instantly now :+1:
